src/d3a/models/area/markets.py

```diff
@@ -15,84 +15,158 @@
+        if ConstSettings.GeneralSettings.ENABLE_SETTLEMENT_MARKETS:
+            return (time_slot < current_time_slot.subtract(
+                hours=ConstSettings.GeneralSettings.MAX_AGE_SETTLEMENT_MARKET_HOURS))
+        else:
+            return time_slot < current_time_slot.subtract(
+                minutes=GlobalConfig.slot_length.total_minutes())
+
+    def rotate_markets(self, current_time: DateTime) -> None:
+        """Deal with market rotation of different types."""
+        self._move_markets_to_past(current_time)
+        self._delete_past_markets(current_time)
+
+        if ConstSettings.BalancingSettings.ENABLE_BALANCING_MARKET:
+            self._move_balancing_markets_to_past_balancing(current_time)
+            self._delete_past_balancing_markets(current_time)
+
+        if ConstSettings.GeneralSettings.ENABLE_SETTLEMENT_MARKETS:
+            self._move_settlement_markets_to_past_settlement(current_time)
+            self._delete_past_settlement_markets(current_time)
+
+        self._update_indexed_future_markets()
+
+    def _move_markets_to_past(self, current_time: DateTime) -> None:
+        """Move slot markets to self.past_markets."""
+        # the conversion to list is needed in order to omit 'mutated during iteration' error
+        for time_slot in list(self.markets.keys()):
+            if time_slot < current_time:
+                market = self.markets.pop(time_slot)
                 market.readonly = True
-                self._delete_past_markets(past_markets)
-                past_markets[timeframe] = market
-                self.log.trace("Moving {t:%H:%M} {m} to past"
-                               .format(t=timeframe, m=past_markets[timeframe].name))
-
-    def _delete_past_markets(self, past_markets):
-        if not d3a.constants.RETAIN_PAST_MARKET_STRATEGIES_STATE:
-            delete_markets = [pm for pm in past_markets if
-                              pm not in self.markets.values()]
-            for pm in delete_markets:
-                if ConstSettings.GeneralSettings.EVENT_DISPATCHING_VIA_REDIS:
-                    past_markets[pm].redis_api.stop()
-                del past_markets[pm].offers
-                del past_markets[pm].trades
-                del past_markets[pm].offer_history
-                del past_markets[pm].notification_listeners
-                del past_markets[pm].bids
-                del past_markets[pm].bid_history
-                del past_markets[pm].traded_energy
-                del past_markets[pm]
+                self.past_markets[time_slot] = market
+                self.log.debug(
+                    f"Moving {self.past_markets[time_slot]} to past.")
+
+    def _move_balancing_markets_to_past_balancing(self, current_time: DateTime) -> None:
+        """Move balancing markets to self.past_balancing_markets."""
+        # the conversion to list is needed in order to omit 'mutated during iteration' error
+        for time_slot in list(self.balancing_markets.keys()):
```

Comment:
```suggestion
        for time_slot in self.balancing_markets.keys():
```

Reply:
@faizan2590 the comment above explains why the `list` conversion is needed in this case.

Reply:
Oh, thanks for pointing out, missed that.